### PR TITLE
spirv-val: Validate vulkan debug info similarly to opencl debug info

### PR DESCRIPTION
**Pull request by greg-lunarg**

### CODE REVIEW

**Comment by greg-lunarg** [comment]
@jaebaek Please begin review. I will be adding some additional regression tests in parallel.

File: source/val/validate_extensions.cpp [hunk 1]
```diff
@@ -68,25 +92,50 @@ spv_result_t ValidateOperandForDebugInfo(
...
+}
+
 #define CHECK_OPERAND(NAME, opcode, index)                                  \
   do {                                                                      \
     auto result = ValidateOperandForDebugInfo(_, NAME, opcode, inst, index, \
                                               ext_inst_name);               \
     if (result != SPV_SUCCESS) return result;                               \
   } while (0)
 
+#define CHECK_LITERAL_OPERAND(NAME, index)                   \
```
Comment (jaebaek) [inline source/val/validate_extensions.cpp:118]: `CHECK_LITERAL_OPERAND` sounds confusing.
How about `CHECK_CONST_UINT_OPERAND`?

**Review by jaebaek — COMMENTED** [review]
Thanks. Mostly looks good but I have some minor comments.
(sorry for pointing very minor things but I hope it helps to improve the code readability)

File: source/val/validate_extensions.cpp [hunk 2]
```diff
@@ -2668,455 +2691,549 @@ spv_result_t ValidateExtInst(ValidationState_t& _, const Instruction* inst) {
...
-              operand->opcode() != SpvOpConstant) {
+              operand->opcode() != SpvOpFunctionParameter) {
             return _.diag(SPV_ERROR_INVALID_DATA, inst)
                    << ext_inst_name() << ": "
                    << "expected operand Variable must be a result id of "
-                      "OpVariable or OpConstant or DebugInfoNone";
+                      "OpVariable or OpFunctionParameter";
           }
+
+          CHECK_DEBUG_OPERAND("Expression", CommonDebugInfoDebugExpression, 7);
```
[the remark this answers is not in the document]
Comment (greg-lunarg) [reply source/val/validate_extensions.cpp:3099]: These don't have to be Const, right? They can be any int scalar value, right?

File: source/val/validate_extensions.cpp [hunk 3]
```diff
@@ -45,6 +47,28 @@ uint32_t GetSizeTBitWidth(const ValidationState_t& _) {
...
+
+  if (type->GetOperandAs<uint32_t>(1) != 32) {
+    return false;
+  }
+
+  if (type->GetOperandAs<uint32_t>(2) != 0) {
+    return false;
+  }
+
+  return true;
```
[the remark this answers is not in the document]
Comment (greg-lunarg) [reply source/val/validate_extensions.cpp:69]: ok

File: source/val/validate_extensions.cpp [hunk 4]
```diff
@@ -2668,455 +2691,549 @@ spv_result_t ValidateExtInst(ValidationState_t& _, const Instruction* inst) {
...
-          if (IsConstIntScalarTypeWith32Or64Bits(_, component_count)) {
-            // TODO: We need a spec discussion for the bindless array.
-            if (!component_count->word(3)) {
-              invalid = true;
+    // Handle any non-common OpenCL insts, then common
+    if (ext_inst_type == SPV_EXT_INST_TYPE_OPENCL_DEBUGINFO_100 &&
+        OpenCLDebugInfo100Instructions(ext_inst_index) ==
+            OpenCLDebugInfo100DebugModuleINTEL) {
+      // No checking required
+    } else {
```
[the remark this answers is not in the document]
Comment (greg-lunarg) [reply source/val/validate_extensions.cpp:2714]: ok

File: source/val/validate_extensions.cpp (hunk 1 above)
Comment (greg-lunarg) [reply source/val/validate_extensions.cpp:118]: ok

File: source/val/validate_extensions.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaebaek) [reply source/val/validate_extensions.cpp:3099]: > These don't have to be Const, right? They can be any int scalar value, right?

Yes.

I just realized that my code to validate `Indexes` operands of `DebugValue` is wrong.
https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/val/validate_extensions.cpp#L3096

Could you please update the `DebugValue` validation as well?
In addition, the [error message](https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/val/validate_extensions.cpp#L3099) is also weird and wrong.

[the remark this answers is not in the document]
Comment (jaebaek) [reply source/val/validate_extensions.cpp:3099]: I mean `Indexes` operands of both `DebugValue` and `DebugDeclare` can be any int scalar value.

**Comment by greg-lunarg** [comment]
@jaebaek I have made the changes you requested. I have also added regression tests for Vulkan.DebugInfo.100 equivalent to what exists for OpenCL.DebugInfo.100. I had to add a few bug fixes to get these to pass. Mostly it was adding logic to deal with IDs of OpConstants rather than literals because of the NonSemantic nature of the Vulkan debug info. Please review.

**Comment by greg-lunarg** [comment]
Note I had to split val_ext_inst_test.cpp and create val_ext_inst_debug_test.cpp which holds all the debugInfo tests. I had to do the split due to object file limits on Windows.

**Review by jaebaek — APPROVED** [review]
Just one last thing .. the debug of _Indexes_ operand check of `CommonDebugInfoDebugValue` https://github.com/KhronosGroup/SPIRV-Tools/pull/4466/files#diff-b4a221138c4c641c6af1de391f9345453052d8377d306708f1d4639650ecf251R3229 is not fixed.

It still checks if _Indexes_ operands consist of constants with int scalar type or `DebugVariable` with int scalar type.
Please fix it or send a separate PR or open a bug for it (if you are busy I will fix it later).

**Comment by jaebaek** [comment]
It must be handled in a way similar to what you do for _Indexes_ operands of `CommonDebugInfoDebugDeclare`:

```
            for (uint32_t word_index = 8; word_index < num_words;
                 ++word_index) {
              auto index_inst = _.FindDef(inst->word(word_index));
              auto type_id = index_inst != nullptr ? index_inst->type_id() : 0;
              if (type_id == 0 || !IsIntScalar(_, type_id, false, false))
                return _.diag(SPV_ERROR_INVALID_DATA, inst)
                       << ext_inst_name() << ": "
                       << "expected index must be scalar integer";
            }
```

**Comment by jaebaek** [comment]
Never mind. I just opened a bug. Let me submit this code.

**Comment by greg-lunarg** [comment]
Thanks! I noticed the DebugValue issue as well and was going to follow up once this was checked in.

**Pull request merged**